pkg/auth/handler/webapp/verify_identity.go

```diff
@@ -116,34 +142,105 @@ func (h *VerifyIdentityHandler) ServeHTTP(w http.ResponseWriter, r *http.Request
 		RedirectURI:     "/verify_identity/success",
 		KeepAfterFinish: true,
 	}
-	intent := intents.NewIntentVerifyIdentityResume(r.Form.Get("id"))
 
-	ctrl.Get(func() error {
-		session, err := ctrl.InteractionSession()
-		if errors.Is(err, webapp.ErrSessionNotFound) ||
-			errors.Is(err, webapp.ErrInvalidSession) {
-			session = nil
+	verificationCodeID := r.Form.Get("id")
+	intent := intents.NewIntentVerifyIdentityResume(verificationCodeID)
+
+	inputFn := func() (input interface{}, err error) {
+		err = VerifyIdentitySchema.PartValidator(VerifyIdentityRequestSchema).ValidateValue(FormToJSON(r.Form))
+		if err != nil {
+			return
 		}
 
-		if session != nil {
-			graph, err := ctrl.InteractionGet()
+		code := r.Form.Get("x_code")
+
+		input = &InputVerificationCode{
+			Code: code,
+		}
+		return
+	}
+
+	ctrl.Get(func() error {
+		if verificationCodeID != "" {
+			// The verification code ID is non-empty.
+			// So this page was opened by verification link.
+			// We have two outcomes.
+			// If the verification is requested by the user, we want to start IntentVerifyIdentityResume.
+			// Otherwise, we want to update the current SessionStep FormData.
+
+			code, err := h.Verifications.GetCode(verificationCodeID)
 			if err != nil {
 				return err
 			}
 
-			data, err := h.GetData(r, w, graph)
-			if err != nil {
+			var graph *interaction.Graph
+			session, err := ctrl.GetSession(code.WebSessionID)
+			if errors.Is(err, webapp.ErrSessionNotFound) || errors.Is(err, webapp.ErrInvalidSession) {
+				session = nil
+			} else if err != nil {
 				return err
 			}
 
-			h.Renderer.RenderHTML(w, r, TemplateWebVerifyIdentityHTML, data)
+			var action string
+			if session == nil {
+				// The web session is invalid.
+				// Here we assume the verification is requested by the user.
+				// This is because if verification is NOT requested by the user, the interaction can never continue.
+				action = VerifyIdentityActionResume
```

Comment:
This logic seems a bit strange? I think this case is when user perform verification as part of authentication, and use the link in email after the interaction expired/completed/invalidated. In that case, would returning a `WebUIInvalidSession` error and render error page more appropriate?

Reply:
This can also be the case where the user requested verification in the settings and wait until the original interaction become expired. If we do not assume VerifyIdentityActionResume here, IntentVerifyIdentityResume is never reachable.

Reply:
I think the root cause is that we can't distinguish whether the verification code comes from authentication flow (non-resumable if invalid) or settings page (resumable if invalid). Maybe we can add a flag to verification code to indicate this? Otherwise in my mentioned case, user may see confusing error when trying to click next?